src/main/java/org/jvnet/hudson/test/JenkinsRule.java

```diff
@@ -792,121 +780,24 @@ public File getWebAppRoot() throws Exception {
-            throws Exception {
-        WebAppContext context = _createWebAppContext2(
-                contextPath,
-                (x) -> localPort = x,
-                getClass().getClassLoader(),
-                localPort,
-                this::configureUserRealm,
-                contextAndServerConsumer);
-        server = context.getServer();
+        JettyProvider jp = findJettyProvider();
+        JettyProvider.Context jpc = jp.createWebServer(localPort, contextPath, this::configureUserRealm);
+        localPort = jpc.localPort();
+        server = jpc.server();
         LOGGER.log(Level.INFO, "Running on {0}", getURL());
-        return context.getServletContext();
+        return jpc.servletContext();
     }
 
-    /**
-     * Creates a web server on which Jenkins can run
-     *
-     * @param contextPath          the context path at which to put Jenkins
-     * @param portSetter           the port on which the server runs will be set using this function
-     * @param classLoader          the class loader for the {@link WebAppContext}
-     * @param localPort            port on which the server runs
-     * @param loginServiceSupplier configures the {@link LoginService} for the instance
-     * @return                     the {@link Server}
-     * @since 2.50
-     */
-    public static WebAppContext _createWebAppContext2(
-            String contextPath,
-            Consumer<Integer> portSetter,
-            ClassLoader classLoader,
-            int localPort,
-            Supplier<LoginService> loginServiceSupplier)
-            throws Exception {
-        return _createWebAppContext2(contextPath, portSetter, classLoader, localPort, loginServiceSupplier, null);
-    }
-    /**
-     * Creates a web server on which Jenkins can run
-     *
-     * @param contextPath              the context path at which to put Jenkins
-     * @param portSetter               the port on which the server runs will be set using this function
-     * @param classLoader              the class loader for the {@link WebAppContext}
-     * @param localPort                port on which the server runs
-     * @param loginServiceSupplier     configures the {@link LoginService} for the instance
-     * @param contextAndServerConsumer configures the {@link WebAppContext} and the {@link Server} for the instance, before they are started
-     * @return                         the {@link Server}
-     * @since 2.50
-     */
-    public static WebAppContext _createWebAppContext2(
-            String contextPath,
-            Consumer<Integer> portSetter,
-            ClassLoader classLoader,
-            int localPort,
-            Supplier<LoginService> loginServiceSupplier,
-            @CheckForNull BiConsumer<WebAppContext, Server> contextAndServerConsumer)
-            throws Exception {
-        QueuedThreadPool qtp = new QueuedThreadPool();
```

Comment:
All of the deleted code in this file was previously duplicated with `HudsonTestCase`. It is now consolidated in:

- `JettyProvider`
- `Jetty12EE9Provider`
- `Jetty12EE10Provider`